lib/models/package-info-cache.js

```diff
@@ -0,0 +1,810 @@
+    }
+
+    return this.addonConstructor;
+  }
+}
+
+class NodeModulesList {
+  constructor(realPath, cache) {
+    this.realPath = realPath;
+    this.cache = cache;
+    this.entries = Object.create(null);
+    this.errors = new ErrorList();
+  }
+
+  _addError(errorType, errorData) {
+    this.errors.addError(errorType, errorData);
+  }
+
+  _insertError(errorEntry) {
+    this.errors.insertError(errorEntry);
+  }
+
+  _hasErrors() {
+    return this.errors.hasErrors();
+  }
+
+  _addEntry(entryName, entryVal) {
+    this.entries[entryName] = entryVal;
+  }
+
+  /**
+   * Given a package name (which may include a scope on the front),
+   * return the PackageInfo that corresponds to the name. 
+   *
+   * Returns null if the package is not found.
+   */
+  _findPackage(packageName) {
+    let val;
+
+    if (packageName.startsWith("@")) {
+      let parts = packageName.split("/");
+      let entry = this.entries[parts[0]]; // scope
+      val =
+        entry instanceof NodeModulesList
+        ? entry._findPackage(parts[1]) // the real name
+        : null;
+    } else {
+      val = this.entries[packageName];
+    }
+
+    return val;
+  }
+}
+
+/**
+ * Class that stores entries that are either PackageInfo or NodeModulesList objects.
+ * The entries are stored in a map keyed by real directory path.
+ */
+class PackageInfoCache {
+  constructor() {
+    this.entries = Object.create(null);
+    this.errors = new ErrorList();
+  }
+
+  _insertError(errorEntry) {
+    this.errors.insertError(errorEntry);
+  }
+
+  /**
+   * Private method to ask just if our PIC-level errors object has errors.
+   */
+  _hasErrors() {
+    return this.errors.hasErrors();
+  }
+
+  /**
+   * Callable externally, does the cache have any objects with errors?
+   */
+  hasErrors() {
+    if (this._hasErrors) {
+      return true;
+    }
+
+    let paths = Object.keys(this.entries);
+
+    if (paths.find(entryPath => this.getEntry(entryPath)._hasErrors())) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Gather all the errors in the PIC and any cached objects, then dump them
+   * out to the console.
+   */
+  showErrors() {
+    this._showObjErrors(this);
+
+    let paths = Object.keys(this.entries).sort();
+
+    paths.forEach(entryPath => {
+      this._showObjErrors(this.getEntry(entryPath));
+    });
+  }
+
+  _showObjErrors(obj) {
+    if (!obj._hasErrors()) {
+      return;
+    }
+
+    console.log("");
+    let rootPath;
+    
+    if (obj instanceof PackageInfoCache) {
+      console.warn("Top level errors:");
+      rootPath = this.realPath;
+    } else {
+      console.warn(obj.realPath + ":");
+      rootPath = obj.realPath;
+    }
+
+    let errorList = obj.errors;
+
+    errorList.getErrors().forEach(errorEntry => {
+      switch(errorEntry.type) {
+      case ERROR_PACKAGE_DIR_MISSING:
+        console.warn(`   Missing package directory at relative path '${path.relative(rootPath, errorEntry.data)}'`);
```

Comment:
currently we use console-ui to interact with the console. I suspect this will also need to.

Reply:
Okay, I'll go figure out how to do that. I was wondering what was needed.